File: backend/app/core/certificates.py
from io import BytesIO

from reportlab.lib import colors
from reportlab.lib.pagesizes import A4, landscape
from reportlab.pdfgen import canvas


def _pdf_text(value: str) -> str:
    return value.encode("latin-1", "replace").decode("latin-1")


def generate_participation_certificate(*, name: str, rider_id: int, route: str) -> bytes:
    buffer = BytesIO()
    page_width, page_height = landscape(A4)
    document = canvas.Canvas(buffer, pagesize=(page_width, page_height))

    document.setFillColor(colors.HexColor("#071313"))
    document.rect(0, 0, page_width, page_height, fill=1, stroke=0)
    document.setStrokeColor(colors.HexColor("#d9ff38"))
    document.setLineWidth(5)
    document.rect(28, 28, page_width - 56, page_height - 56, fill=0, stroke=1)

    document.setFillColor(colors.HexColor("#d9ff38"))
    document.setFont("Helvetica-Bold", 15)
    document.drawCentredString(page_width / 2, page_height - 92, "NV CYCLOTHON 2026")

    document.setFillColor(colors.white)
    document.setFont("Helvetica-Bold", 36)
    document.drawCentredString(page_width / 2, page_height - 150, "PARTICIPATION CERTIFICATE")

    document.setFont("Helvetica", 16)
    document.drawCentredString(page_width / 2, page_height - 202, "This certificate is proudly presented to")
    document.setFillColor(colors.HexColor("#d9ff38"))
    document.setFont("Helvetica-Bold", 31)
    document.drawCentredString(page_width / 2, page_height - 252, _pdf_text(name))

    document.setFillColor(colors.white)
    document.setFont("Helvetica", 15)
    document.drawCentredString(
        page_width / 2,
        page_height - 305,
        f"for participating in the {_pdf_text(route)} bicycle route",
    )
    document.drawCentredString(
        page_width / 2,
        page_height - 332,
        "NV Cyclothon · 18 October 2026 · Rewa, Madhya Pradesh",
    )

    document.setFillColor(colors.HexColor("#ff5f3d"))
    document.setFont("Helvetica-Bold", 13)
    document.drawCentredString(page_width / 2, 95, f"RIDER ID  #{rider_id}")
    document.setFillColor(colors.white)
    document.setFont("Helvetica", 11)
    document.drawCentredString(page_width / 2, 70, "With appreciation from NV Cyclothon and Rewa Cycling Federation")

    document.showPage()
    document.save()
    return buffer.getvalue()

File: backend/src/services/certificates.js
const PDFDocument = require("pdfkit");

function generateParticipationCertificate({ name, riderId, route }) {
  const document = new PDFDocument({ size: "A4", layout: "landscape", margin: 0 });
  const chunks = [];

  return new Promise((resolve) => {
    document.on("data", (chunk) => chunks.push(chunk));
    document.on("end", () => resolve(Buffer.concat(chunks)));

    const width = document.page.width;
    const height = document.page.height;

    document.rect(0, 0, width, height).fill("#071313");
    document.lineWidth(5).strokeColor("#d9ff38").rect(28, 28, width - 56, height - 56).stroke();

    document.fillColor("#d9ff38").font("Helvetica-Bold").fontSize(15).text("NV CYCLOTHON 2026", 0, 78, {
      align: "center",
    });

    document.fillColor("white").fontSize(36).text("PARTICIPATION CERTIFICATE", 0, 120, {
      align: "center",
    });

    document.font("Helvetica").fontSize(16).text("This certificate is proudly presented to", 0, 190, {
      align: "center",
    });

    document.fillColor("#d9ff38").font("Helvetica-Bold").fontSize(31).text(String(name), 0, 230, {
      align: "center",
    });

    document.fillColor("white").font("Helvetica").fontSize(15).text(
      `for participating in the ${String(route)} bicycle route`,
      0,
      300,
      { align: "center" }
    );
    document.text("NV Cyclothon - 18 October 2026 - Rewa, Madhya Pradesh", 0, 328, {
      align: "center",
    });

    document.fillColor("#ff5f3d").font("Helvetica-Bold").fontSize(13).text(`RIDER ID  #${riderId}`, 0, height - 110, {
      align: "center",
    });

    document.fillColor("white").font("Helvetica").fontSize(11).text(
      "With appreciation from NV Cyclothon and Rewa Cycling Federation",
      0,
      height - 86,
      { align: "center" }
    );

    document.end();
  });
}

module.exports = {
  generateParticipationCertificate,
};
